### PR TITLE
Fix linkage for `cdef public` functions in C++ mode

### LINKED ISSUE #1839: Functions with C++ return type are `extern "C"`
Thats a C++ compiler error.
All functions with C++ in their signatures should be `extern "C++"`.

### PR DESCRIPTION
**Pull request by EpigeneMax**
Fixes #1839 

`cdef public` functions should be declared with the appropriate linkage:
- in C mode, either `extern` or `extern "C"`, depending on whether the header file is included in (resp. object code is linked against) a C or a C++ compilation unit. Choice is made at compile-time through `#ifdef __cplusplus` macros. NB: This is the current behavior.
- in C++ mode, `extern "C++"` is the only option, as C code cannot call C++ code. Note that `extern "C++"` should be preferred over `extern` to allow users to `#include` the C++ header inside a `extern "C"` block (which is legal, although barely used).

Note that the current behavior is OK for C mode, but is incorrect for the C++ mode. As described in #1839, this incorrect behavior is diagnosed by compilers emitting warnings when `cdef public` functions return a C++ type (e.g. `std::vector`).
The test introduced in this PR checks that the current behavior for C mode (with both C and C++ compatibility) is preserved, and that the behavior for C++ mode is fixed.

### CODE REVIEW

**Comment by da-woods** [comment]
I'm not completely 100% convinced that this is right, but it probably is.

If you have a public function with argument types that doesn't need C++ (e.g. `int func(int)`) then this change means it can't be called from C. Obviously if you have argument types that do need C++ then it's currently broken. I don't think there's anything we can do to tell that easily when we compile it though (a struct could be either, for example).

It slightly depends on whether you're viewing it as "Cython module implemented in C++" or "Cython module with C++ public interface". 

On balance I think the change is probably reasonable - it'd be easy for a user to write a very thin wrapper to declare `extern "C"` functions if they wanted to anyway.

**Comment by da-woods** [comment]
I've thought about it more and believe this is fine. See https://github.com/cython/cython/issues/1839#issuecomment-1254169925. The user can always manually override the `__PYX_EXTERN_C` symbol. So defaulting to C++ seems right...

**Comment by EpigeneMax** [comment]
I have updated the test to account for the scenario where `__PYX_EXTERN_C` is defined by the end-user, and to remove clang-specific warnings.
I think we are good for review.

There is one failing check though, but it seems it ran into a problem during environment installation... Maybe it should needs to be re-run.

File: tests/run/cpp_extern.srctree [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+size_t size_vector1() {
+    return get_vector().size();
+}
+
+######## foo2.cpp ########
+
+#include <vector>
+extern "C" {
+#include "foo.h"
+}
```
Comment (da-woods) [inline tests/run/cpp_extern.srctree:72]: What's the `extern "C"` here testing? I'm just wondering if it's really a C language feature and perhaps isn't something that Cython needs to test

Comment (EpigeneMax) [reply tests/run/cpp_extern.srctree:72]: It tests that the `cdef public` functions are flagged as `extern "C++"` (and not `extern`) even when in a C++ compilation unit.
The rationale is that it is legal to `#include` a header from within a `extern "C"` block. In this scenario, all declarations of the header are seen with C-linkage, except those explicitly flagged `extern "C++"`.

This usage of `extern "C"` `#include` is probably a niche, yet it is legal, so I felt that Cython-generated code should comply with it, and compliance should be tested.

Comment (da-woods) [reply tests/run/cpp_extern.srctree:72]: Fair enough - that makes sense. It seems worth a comment since it isn't obvious at first look (at least to me)

Comment (EpigeneMax) [reply tests/run/cpp_extern.srctree]: Comment added in commit [79f8003](https://github.com/cython/cython/pull/5040/commits/79f800360b25b8820329627d992b4b777daabdae)

File: Cython/Compiler/ModuleNode.py [hunk 2]
```diff
@@ -876,14 +876,17 @@ def generate_module_preamble(self, env, options, cimported_modules, metadata, co
...
-    def generate_extern_c_macro_definition(self, code):
+    def generate_extern_c_macro_definition(self, code, env):
         name = Naming.extern_c_macro
         code.putln("#ifndef %s" % name)
-        code.putln("  #ifdef __cplusplus")
-        code.putln('    #define %s extern "C"' % name)
-        code.putln("  #else")
-        code.putln("    #define %s extern" % name)
-        code.putln("  #endif")
+        if env.is_cpp():
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py:882]: If this is the only functionality that we need from env, then I'd rather add an argument `is_cpp=False` than passing  all of`env` into this method.

File: Cython/Compiler/ModuleNode.py [hunk 3]
```diff
@@ -262,7 +262,7 @@ def h_entries(entries, api=0, pxd=0):
             api_guard = self.api_name(Naming.api_guard_prefix, env)
             h_code_start.putln("#ifndef %s" % api_guard)
             h_code_start.putln("")
-            self.generate_extern_c_macro_definition(h_code_start)
+            self.generate_extern_c_macro_definition(h_code_start, env)
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py:265]: I think (!) the header file isn't necessarily C++ specific even when we generate C++ code, so this should better not look at `env.is_cpp()` but use pure compile-time C/C++ features.

File: docs/examples/tutorial/embedding/embedded.pyx [hunk 4]
```diff
@@ -3,6 +3,7 @@
 # The following two lines are for test purposes only, please ignore them.
 # distutils: sources = embedded_main.c
 # tag: py3only
+# tag: no-cpp
```
Comment (scoder) [inline docs/examples/tutorial/embedding/embedded.pyx:6]: This seems an unrelated change that shouldn't be included in this PR.

File: tests/compile/excvalcheck.h [hunk 5]
```diff
@@ -1,12 +1,6 @@
-#ifdef __cplusplus
-extern "C" {
-#endif
```
Comment (scoder) [inline tests/compile/excvalcheck.h:3]: This is in test code. Why did you remove these?

**Review by scoder — COMMENTED** [review]
Nice test that you added there.

File: Cython/Compiler/ModuleNode.py (hunk 3 above)
Comment (da-woods) [reply Cython/Compiler/ModuleNode.py:265]: My view was:
* if you're passing any C++ class (and a struct can be a hidden a C++ class) then the header is C++ specific. It's definitely very hard for Cython to tell if it's C++ specific.
* The user can always redefine the `__PYX_EXTERN_C` macro themselves, so we aren't forcing them into being C++ specific.
* With that said, I'm wondering what's wrong with declaring the functions as plain `extern` in both cases. There possibly is something I'm missing though

File: Cython/Compiler/ModuleNode.py (hunk 2 above)
Comment (EpigeneMax) [reply Cython/Compiler/ModuleNode.py:882]: Agreed.

File: docs/examples/tutorial/embedding/embedded.pyx (hunk 4 above)
Comment (EpigeneMax) [reply docs/examples/tutorial/embedding/embedded.pyx:6]: The intention is to exclude this test from C++ backend runs.
Running this test with the C++ backend fails: `.pyx` file is cythonized to C++, and the generated C++ header is included in a `.c` compilation unit, and finally the compiler rightfully rejects the `extern "C++"` declaration.

If we want to keep it in the C++ backend, then the proper solution would be to redefine the `__PYX_EXTERN_C` macro.

File: tests/compile/excvalcheck.h (hunk 5 above)
Comment (EpigeneMax) [reply tests/compile/excvalcheck.h:3]: I am just adjusting existing tests to the new behavior.
With the old behavior, functions were declared as `extern` in C mode and `extern "C"` in C++ mode by Cython, and all declarations within a compilation unit must agree.
With the new behavior, functions are declared as `extern` in C mode and `extern "C++"` in C++ mode by Cython, so these `extern "C"` blocks make the compilation fail.

**Comment by EpigeneMax** [comment]
Any update on this?
@scoder Did you have time to review my comments? Or did I miss something?

**Comment by da-woods** [comment]
Thanks @EpigeneMax

**Comment by da-woods** [comment]
Need to look into this but the test is breaking on windows because the command line argument `"-D__PYX_EXTERN_C=extern "C""` isn't escaped properly
